automation/Extension/firefox/lib/http-post-parser.js

```diff
@@ -0,0 +1,250 @@
+// Incorporates code from: https://github.com/redline13/selenium-jmeter/blob/6966d4b326cd78261e31e6e317076569051cac37/content/library/recorder/HttpPostParser.js
+
+const {Cc, Ci, CC, Cu, components} = require("chrome");
+var loggingDB     = require("./loggingdb.js");
+
+var HttpPostParser = function(stream) {
+  // Scriptable Stream Constants
+  this.seekablestream = stream;
+  this.stream = components.classes["@mozilla.org/scriptableinputstream;1"].createInstance(components.interfaces.nsIScriptableInputStream);
+  this.stream.init(this.seekablestream);
+
+  this.postBody = "";
+  this.postLines = [];
+  this.postHeaders = [];
+  // Check if the stream has headers
+  this.hasheaders = false;
+  this.body = 0;
+  if (this.seekablestream instanceof components.interfaces.nsIMIMEInputStream) {
+    this.seekablestream.QueryInterface(components.interfaces.nsIMIMEInputStream);
+    this.hasheaders = true;
+    this.body = -1;
+  } else if (this.seekablestream instanceof components.interfaces.nsIStringInputStream) {
+    this.seekablestream.QueryInterface(components.interfaces.nsIStringInputStream);
+    this.hasheaders = true;
+    this.body = -1;
+  }
+}
+
+HttpPostParser.prototype.rewind = function() {
+  this.seekablestream.seek(0, 0);
+};
+
+HttpPostParser.prototype.tell = function() {
+  return this.seekablestream.tell();
+};
+
+HttpPostParser.prototype.readLine = function() {
+  var line = "";
+  var size = this.stream.available();
+  for (var i = 0; i < size; i++) {
+    var c = this.stream.read(1);
+    if (c == '\r') {
+    } else if (c == '\n') {
+      break;
+    } else {
+      line += c;
+    }
+  }
+  return line;
+};
+
+// visitor can be null, function has side-effect of setting body
+HttpPostParser.prototype.headers = function() {
+  if (this.hasheaders) {
+    this.rewind();
+    var line = this.readLine();
+    while (line) {
+      var keyValue = line.match(/^([^:]+):\s?(.*)/);
+      // match can return null...
+      if (keyValue) {
+        this.postHeaders[keyValue[1]] = keyValue[2];
+      } else {
+        this.postLines.push(line);
+      }
+      line = this.readLine();
+    }
+    this.body = this.tell();
+  }
```

Comment:
It was very confusing to me what headers you are parsing here. I originally thought the stream you were accessing contained the entire request and thus you needed to parse the HTTP request headers as well as the post body content. It wasn’t clear to me why some request streams might not have headers and why there was any uncertainty in extracting them. It wasn’t until I walked through the `parsePostRequest` method that I found the link to the stack overflow explaining post headers (line 134 of `http-post-parser.js`). Can we add a comment to this method that explains how these are distinct from request headers and why there is ambiguity in parsing them. A simple example request might be helpful to understand HTTP request headers, post-body headers, and headers within multipart post-bodies.